Add support for various deflate compression levels

soulseekah opened the pull request:
Different compression levels yield a specific second byte in the magic
header for the deflate encoding. All of them can be decoded by the same
functions without any issues. Let them through, as they've been seen in
the wild.

Fixes #301

rmccue commented on library/Requests.php:
```diff
@@ -821,7 +821,9 @@ public static function flattern($array) {
-		if (substr($data, 0, 2) !== "\x1f\x8b" && substr($data, 0, 2) !== "\x78\x9c") {
+		// All valid deflate, gzip header magic markers
+		$valid_magic = array("\x1f\x8b", "\x78\x01", "\x78\x5e", "\x78\x9c", "\x78\xda");
+		if (!in_array(substr($data, 0, 2), $valid_magic)) {
```
This needs to use the `$strict` parameter to `in_array`, but otherwise looks good.

soulseekah replied:
Fixed in follow up commit. Tests included.

rmccue reviewed (dismissed):
This also needs tests in the [Encoding.php](https://github.com/rmccue/Requests/blob/master/tests/Encoding.php) suite.

soulseekah commented:
Added strict in_array checking, added compression level tests (generated using zpipe.c from the zlib 1.2.11 using levels 1, 3 and 9, 5 was already being tested as the generic case). Also added Encoding.php to the test suite, it was not explicitly defined in phpunit.xml.dist and thus was not being run at all.

jrfnl commented:
@schlessera I've just been looking at this PR. This seems like a good change. What do you think ?

To get this ready to merge, IMO, the following actions are needed:
* Rebase the PR.
    This should automatically remove the change to `tests/phpunit.xml.dist`. If not, that change needs to be removed as this is now fixed more thoroughly already.
* Move the `$valid_magic` array to a class constant and document the source of the magic markers.
    Nice to have/micro-optimization: Use the markers as the keys in the array instead of as the values and use `isset()` instead of `in_array()`.

jrfnl commented:
I've made the changes as per the above comment. This PR is now ready for review/merge.

schlessera commented on src/Requests.php:
```diff
@@ -184,6 +184,25 @@ class Requests {
+	/**
+	 * All (known) valid deflate, gzip header magic markers.
+	 *
+	 * These markers related to different compression levels.
```
```suggestion
	 * These markers relate to different compression levels.
```

soulseekah commented on src/Requests.php:
```diff
@@ -187,7 +187,7 @@ class Requests {
 	 * All (known) valid deflate, gzip header magic markers.
 	 *
-	 * These markers related to different compression levels.
+	 * These markers relate to different compression levels.
```
Thanks :)

schlessera commented:
Thanks for the PR, @soulseekah !
